mkosi/backend.py

```diff
@@ -519,6 +520,29 @@ def run(
         die(f"{cmdline[0]} not found in PATH.")
 
 
+def run_with_backoff(
+    cmdline: Sequence[PathString],
+    check: bool = True,
+    delay_interrupt: bool = True,
+    stdout: _FILE = None,
+    stderr: _FILE = None,
+    *,
+    attempts: int,
+    **kwargs: Any,
+) -> CompletedProcess:
+    delay = 0.0
+    for attempt in range(attempts):
+        try:
+            return run(cmdline, check, delay_interrupt, stdout, stderr, **kwargs)
+        except subprocess.CalledProcessError:
+            if attempt == attempts - 1:
+                raise
+            time.sleep(delay)
+            delay = min(delay * 2 + 0.01, 1)
+
+    assert False  # make mypy happy
```

Comment:
What about the following instead:

```
for attempt in range(attempts):
    try:
        return run(cmdline, check, delay_interrupt, stdout, stderr, **kwargs)
    except subprocess.CalledProcessError as e:
        time.sleep(delay)
        delay = min(delay * 2 + 0.01, 1)

raise e
```

Reply:
`e` doesn't leak out of the except scope:
```
In [1]: try:
   ...:     1/0
   ...: except ZeroDivisionError as e:
   ...:     print("oops")
   ...: 
oops

In [2]: e
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-3-094e3afb2fe8> in <module>
----> 1 e

NameError: name 'e' is not defined
```

Reply:
Damn I was hoping it would, nvm then

Reply:
It could be saved, e.g. with `ee = e`, but the bigger problem is that this sleep one extra time. And since the backoff is exponential, we essentially sleep the longest at the end.